Writing custom check functions vignette (#127)
hubverse-org/hubValidations: vignettes/articles/writing-custom-fns.Rmd

# Anatomy of a check function

While source code of existing `hubValidations` `check_*()` functions can be a good place to start when writing custom check functions, it is important to understand the structure of a check function, particularly the expected inputs and outputs.

At it's most basic, a custom check function should:

- evaluate whether a condition is met
- return an appropriate check condition object

In addition, if the check condition is not met, it's also helpful to capture any details that can guide users towards specifics of the failure and how to fix it

In general, `hubValidations` check functions evaluate conditions with respect to one or more of the following:

- Model output submission files
- Model output submission file content (i.e data)
- Model metadata files

## `create_custom_check()` for creating custom check function templates

To help you get started on the right path, we also provide function `create_custom_check()` for creating a basic custom check function from a template. 

> The function requires a name for the new custom check function, creates a `.R` script file in `src/validations/R` directory in the root of the hub, the recommended location. The script contains basic skeleton code to create a named custom check function. 

Author: Edited the suggestions and added an example for clarity in c1877672934b8395ed57d6ab61a6f8675eff9089

> The output of `create_custom_check()` can also be parametarised through a number of arguments to include additional template code snippets.

Author: Done in 103bbd414d97162d5a8046890820f6d3a8909933

- `check_submission_*` for checks that operate on high level properties of the submission (e.g. timing).
- `check_config_*` for checks that operate on the hub configuration files.
- `check_valid_*` for checks that don't fit into the above categories.

To distinguish `hubValidations` package functions from custom ones, we also recommend prefixing custom function names with an additional prefix, e.g. `cstm_` or `cs`.

</div>

## Arguments to custom check functions

### Arguments available in the caller environment 

The `hubValidations` package automatically passes a set of standard objects to custom check functions. These objects are passed to arguments of custom check functions with the same name and do not need to be included in the `args` configuration in the `validations.yml` file.

```{r child="children/_custom-fn-available-args.Rmd", echo=FALSE, results="asis"}
```

### Additional arguments

You can add additional arguments to custom check functions and pass values to them by including them in the `args` configuration in the `validations.yml` file. These values are passed to the custom check function by `hubValidations` when the function is called.

> If you do add additional arguments to a custom check function, you should also add input  checks at the start of the function to ensure inputs are valid. Package `checkmate` contains a wide range of functions for checking inputs.

Reviewer: Maybe we can add a link to the checkmate package here

For example, the optional check `opt_check_tbl_col_timediff()` (which is deployed in exactly the same fashion as custom functions, i.e. through the `validations.yml` file) takes additional arguments `t0_colname`, `t1_colname` and `timediff`. 

```{r}
opt_check_tbl_col_timediff
```

You can add an example extra argument with `extra_args = TRUE` when creating the custom check function with `create_custom_check()`.

```{r}
create_custom_check("cstm_check_tbl_args",
  hub_path = hub_path,
  extra_args = TRUE
)
```

This adds an extra argument `extra_arg` to the custom check function formals as well as an example input check.

> s

Reviewer: Small typo, a `s` is alone here

```{r, code=readLines(r_path(hub_path, "cstm_check_tbl_args")), eval=FALSE}
```

## Function output

### Capturing and returning check results

**Returning a check condition with `capture_check_cnd()`**

> The `capture_check_cnd()` function is used to return a check condition and it's output should also be what a custom check function returns in most cases. The function returns a `<hub_check>` class object depending on the value passed to the `check` argument, which represents the summary of the condition being checked by a given validation function. 

Reviewer: I am not sure what the "most cases" mean here, would it be possible to add more information, please

Author: The only other case is when a function might be skipped and return a `capture_check_info()` which is discussed immediately below

> The `capture_check_cnd()` function is used to return a check condition and it's output should also be what a custom check function returns in most cases. The function returns a `<hub_check>` class object depending on the value passed to the `check` argument, which represents the summary of the condition being checked by a given validation function. 

Author: I'm keeping the "n most cases" because the following section clearly describes a situation where the output of `capture_check_cnd()` is not what the function returns

> The `capture_check_cnd()` function is used to return a check condition and it's output should also be what a custom check function returns in most cases. The function returns a `<hub_check>` class object depending on the value passed to the `check` argument, which represents the summary of the condition being checked by a given validation function. 

Reviewer: I am ok with keeping the `in most cases`, I was just not clear what it meant. The sentence was not clear to me, but I like the new version, it's all clear to me now! thank you again

If the value passed to `check ` is `TRUE`, the function returns a `<message/check_success>` class object. 

If the value is `FALSE`, the output depends on the `error` argument.  

- If `error` is `TRUE`, the function returns a `<message/check_error>` class object. If `error` is `FALSE`, the function returns a `<error/check_failure>` class object, which indicates the check has failed. 

> - If `error` is `TRUE`, the function returns a `<error/check_error>` class object, which indicates the check has failed and additionally causes execution of further custom validation functions to halt. Set `error = TRUE` if downstream checks cannot be run if the current check fails.

Reviewer: I might open an issue for my question, I don't think here is the best place, but I thought we also had a warning option? If not, should we add one? It it might be useful to returns that information for an issue that is not technically incorrect but might be worth checking. For example, we might have a cumulative case projection bigger than the population size, it might be correct due to re-infection but it might be worth checking to be sure, so it might be good to returns a warning in that case, does that make sense?

Reviewer: Thank you for the information, and I apologize I miss that one.  
I have a small issue with the warning being a failure because it technically did not fail, but it's more a philosophical question. To continue my example, if I want to be able to tag projection of cases that are higher than population size. It can be technically correct as people can be re-infected but it will be a surprising projections (in this context) so I want to be able to tag it so the team can easily check that it's what they want to submits.

Reviewer 2: > I want to be able to tag it so the team can easily check that it's what they want to submits.

This is what `<error/check_failure>` (what used to be `<warning/check_failure>` is doing). In the old version of hubValidations, the `<warning/check_failure>` still gave an `Error` in the overall validation result. 

It mostly boils down to semantics. **Regardless of the name (warning or error), they both resulted in an overall validation that did not pass**. 

Since this requires human intervention, if this is a custom validation function, then you can add a bold comment that says "projected cases (n) higher than population size (N) **NOTE: reinfections can cause projections to be higher than the population size**, if this is the case, then the error is a false-positive". Then you as the admin can confirm the following and merge the pull request, overriding the failed check:

1. this is the only check that fails
2. you and the modeler are certain that reinfections are the cause

Reviewer: thanks for the clarification! that make sense, thanks. I think I just worry that people misunderstood the error as something that need to be fixed and is incorrect. But, I guess we can put information in the resulting message to clarify that.

Author: @LucieContamin`warning/check_failure>`s have always been considered failures that need fixing in `hubValidations` and would cause PR workflows to fail (as per @zkamvar's comment). Really what you are describing is a completely new feature in `hubValidations` (i.e. something that is not quite a pass but is not a failure ether).

If this is something you might indeed use, I suggest you open an issue about it describing how see the feature being used. In fact, the changes made in 0.6.0 actually allow for the use of warnings for such a purpose now.

> If the value is `FALSE`, the output depends on the `error` argument.  
> 
> - If `error` is `TRUE`, the function returns a `<message/check_error>` class object. If `error` is `FALSE`, the function returns a `<error/check_failure>` class object, which indicates the check has failed. 
> - If `error` is `TRUE`, the function returns a `<error/check_error>` class object, which indicates the check has failed and additionally causes execution of further custom validation functions to halt. Set `error = TRUE` if downstream checks cannot be run if the current check fails.
> 
> The default value of `error` is `FALSE` in `create_custom_check()` but setting it to `TRUE` sets the error argument to `TRUE`.

Reviewer 2: ```suggestion
If the value is `FALSE`, the output depends on the `error` argument.  

- If `error` is `FALSE` (the default), the function returns a `<error/check_failure>` class object, which indicates the check has failed.
- If `error` is `TRUE`, the function returns a `<error/check_error>` class object, which indicates the check has failed and additionally causes execution of further custom validation functions to halt. Set `error = TRUE` if downstream checks cannot be run if the current check fails.
```

```{r}
create_custom_check("cstm_check_tbl_error",
  hub_path = hub_path, error = TRUE
)
```

```{r, code=readLines(r_path(hub_path, "cstm_check_tbl_error")), eval=FALSE}
```

### Skipping a check

> **Returning a check message with `capture_check_info()`**

Reviewer 2: ```suggestion
```

Author: I don't really want it as a sub heading as it's not adding anything to the organisation tree of the document. It's more of a subtitle for that particular section.

I think it looks fine in the rendered version?
<img width="819" alt="image" src="https://github.com/user-attachments/assets/31ec3950-684c-4768-b23a-69e7bfe2162a">

There's also exactly the same in the previous section:

<img width="830" alt="image" src="https://github.com/user-attachments/assets/7db013ca-7b75-4292-b3e6-053f8dfd24c3">

Sometimes a check function might not always be applicable and a pre-condition needs to be met before the main check itself is performed. **If the pre-condition is not met, the check is usually skipped**. 

**For such checks, the function should return a `<message/check_info>` object**, generated by the `capture_check_info()` function. Use the `msg` argument to explain that a check was skipped and why.

```{r}
capture_check_info(
  "modelA-teamA/2024-09-12-modelA-teamA",
  "Condition for running this check was not met. Skipped."
)
```

For example, the `check_tbl_value_col_ascending()` check function which validates that values are ascending when arranged by increasing `output_type_id` order is only applicable to `cdf` and `quantile` output types. Before proceeding with the main check, the function first checks whether the model output `tbl` contains data for `cdf` and `quantile` output types. If not, the check is skipped.

> ```{r}

Reviewer 2: ````suggestion
```{r, comment = NA}
````

check_tbl_value_col_ascending
```

You can add a pre-condition check block of code with argument `conditional = TRUE` when creating the custom check function with `create_custom_check()`.

```{r}
create_custom_check("cstm_check_tbl_skip",
  hub_path = hub_path,
  conditional = TRUE
)
```

```{r, code=readLines(r_path(hub_path, "cstm_check_tbl_skip")), eval=FALSE}
```

## Loading config files

> Many check conditions against information stored in hub configuration files and these are often read in at the start of the custom check function.

Reviewer: I found this sentence confusing. 
Is "check conditions" the object rendered by function or is it "Many people wants to check (as in validate) conditions against ..." ? if it's the first one, are we missing a verb??

Reviewer 2: ```suggestion
Many checks are conditioned against information stored in hub configuration files and these are often read in at the start of the custom check function.
```

- `cstm_check_tbl_*` for checks that operate on a tibble of model output data (i.e. the contents of a file).
- `cstm_check_meta_*` for checks that operate on model metadata files.
- `cstm_check_submission_*` for checks that operate on high level properties of the submission (e.g. timing).
- `cstm_check_config_*` for checks that operate on the hub configuration files.
- `cstm_check_valid_*` for checks that don't fit into the above categories.

Neither of these recommendations are required for custom functions to work, but consistency is an important aspect of maintaining a hub. 

</div>

## Function inputs / arguments

The minimum inputs required by a custom check function depend on the type of check being performed.

- `file_path`: the relative path to the submission file being validated is required for all check functions. **`file_path` must therefore be included as an argument** in all custom check function.
- `tbl` or `tbl_chr`: a tibble representation of the contents of a model output submission (with column data types matching the hub schema (`tbl`) or an all character version (`tbl_chr`) is also required by any checks that operate on the data in the submission file.

> Since `file_path` and `tbl` are the most common inputs to check functions, `create_custom_check()` includes them as arguments by default. This means that the custom check function will include these objects in the function call environment by default. 

Reviewer 2: This really makes things clear!

> <div class="alert alert-warn">

Author: I've added it but it doesn't seem to do much 🤷‍♀️

> <div class="alert alert-warn">

Author: FIxed thanks!

> 

Reviewer 2: I really like this discussion and I feel like a summary heading here can help drive home the point. 

```suggestion

### Use `tbl` only with `validate_model_data()`

```

</div>

In addition to these, custom check functions can also have additional arguments for inputs required by the check. Some of these inputs **are available in the check caller environment** and can be passed automatically to a custom check function by including an argument with the same name as the input object required in the custom function formals. Other inputs can be passed explicitly to function arguments through [the functions `args` field](deploying-custom-functions#validations-yml-structure) when configuring the `validations.yml` file.

### Arguments available in the caller environment 

```{r child="children/_custom-fn-available-args.Rmd", echo=FALSE, results="asis"}
```

<div class="alert alert-info">

> Note that, when writing custom functions, these objects **do not all need to be specified as arguments** in the function definition. **Only the ones that your custom function actually requires as inputs**.

Reviewer 2: This is short, sweet, and to the point. Love to see it.

If you do add additional arguments to a custom check function, you should also add input  checks at the start of the function to ensure inputs are valid. [The `checkmate` package](https://mllg.github.io/checkmate/) contains a wide range of functions for checking inputs.

For example, the optional check `opt_check_tbl_col_timediff()` (which is deployed in exactly the same fashion as custom functions, i.e. through the `validations.yml` file) takes additional arguments `t0_colname`, `t1_colname` and `timediff`. 

```{r, comment = NA}
opt_check_tbl_col_timediff
```

You can add an example extra argument with `extra_args = TRUE` when creating the custom check function with `create_custom_check()`.

```{r}
create_custom_check("cstm_check_tbl_args",
  hub_path = hub_path,
  extra_args = TRUE
)
```

This adds an extra example argument `extra_arg` to the custom check function formals as well as an example input check to the top of the function body.

```{r, code=readLines(r_path(hub_path, "cstm_check_tbl_args")), eval=FALSE, comment=NA}
```

## Function output

### Capturing and returning check results

> **Returning a check condition with `capture_check_cnd()`**

Reviewer 2: This looks like it's meant to be a heading, so it should be coded as such. 

```suggestion
### Returning a check condition with `capture_check_cnd()`
```

Author: It's more of a subtitle really than a heading
